receiver/receivercreator/rules.go

```diff
@@ -15,10 +15,79 @@
 package receivercreator
 
 import (
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/antonmedv/expr"
+	"github.com/antonmedv/expr/vm"
+
 	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer"
 )
 
-func ruleMatches(rule string, e observer.Endpoint) bool {
-	// TODO
-	return rule == "enabled"
+// rule wraps expr rule for later evaluation.
+type rule struct {
+	program *vm.Program
+}
+
+var ruleRe = regexp.MustCompile("^type.(Pod|Port)")
+
+// newRule creates a new rule instance.
+func newRule(ruleStr string) (rule, error) {
+	if ruleStr == "" {
+		return rule{}, errors.New("rule cannot be empty")
+	}
+	if !ruleRe.MatchString(ruleStr) {
+		// TODO: Try validating against bytecode instead.
+		return rule{}, errors.New("rule must specify type")
+	}
+
+	// TODO: Maybe use https://godoc.org/github.com/antonmedv/expr#Env in type checking
+	// depending on type == specified.
+	v, err := expr.Compile(ruleStr)
+	if err != nil {
+		return rule{}, err
+	}
+	return rule{v}, nil
+}
+
+// eval the rule against the given endpoint.
+func (r *rule) eval(endpoint observer.Endpoint) (bool, error) {
+	var env map[string]interface{}
+
+	ruleTypes := struct {
+		Port bool
+		Pod  bool
+	}{}
+
+	switch o := endpoint.Details.(type) {
+	case observer.Pod:
+		ruleTypes.Pod = true
+		env = map[string]interface{}{
+			"type":   ruleTypes,
+			"name":   o.Name,
+			"labels": o.Labels,
+		}
+	case observer.Port:
+		ruleTypes.Port = true
+		env = map[string]interface{}{
+			"type":     ruleTypes,
+			"name":     o.Name,
+			"port":     o.Port,
+			"pod_name": o.Pod.Name,
```

Comment:
Should we use "pod.name" to match Otel conventions? Or it is not allowed as a variable name in `expr`?

Reply:
This seems incorrect to me? Should we populate pod_name and pod labels if endpoint is `observer.Pod` instead? If this code is correct then it requires explanation since I do not understand it.

Reply:
Yeah you can't use `.` in a variable name but I can still change it so that `pod.name` works. 

The reason labels and pod_name are there is because you may want to use them in your port check. For example:

`type.Port && name == "http" && pod.labels["app"] == "http"`

Reply:
It would be great to use a dot in "pod.name". I think it will make it consistent with "pod.labels".

Reply:
Yeah there's `pod.name` and `pod.labels` now. There's still `name` which is the name of the port.